bokehjs/src/lib/models/ranges/factor_range.ts

```diff
@@ -27,84 +27,75 @@ export type L3OffsetFactor = [string, string, string, number]
 
 export type OffsetFactor = L1OffsetFactor | L2OffsetFactor | L3OffsetFactor
 
-export type L1Mapping = {[key: string]: {value: number}}
-export type L2Mapping = {[key: string]: {value: number, mapping: L1Mapping}}
-export type L3Mapping = {[key: string]: {value: number, mapping: L2Mapping}}
+export type L1Mapping = Map<string, {value: number}>
+export type L2Mapping = Map<string, {value: number, mapping: L1Mapping}>
+export type L3Mapping = Map<string, {value: number, mapping: L2Mapping}>
+
+export type Mapping = L1Mapping | L2Mapping | L3Mapping
 
 export function map_one_level(factors: L1Factor[], padding: number, offset: number = 0): [L1Mapping, number] {
-  const mapping: L1Mapping = {}
+  const mapping: L1Mapping = new Map()
 
   for (let i = 0; i < factors.length; i++) {
     const factor = factors[i]
-    if (factor in mapping)
-      throw new Error(`duplicate factor or subfactor: ${factor}`)
+    if (!mapping.has(factor))
```

Comment:
why add the negation and swap the branch order? It adds cognitive load but I don't see any offsetting benefit

Reply:
My attitude is that the common case comes first and the uncommon later. There may be also a case for performance optimization, but that's irrelevant at this level of abstraction.

Reply:
Interesting, besides the simpler conditional, my mind definitely prefers "check and bail early" 
```
if (mapping.has(factor)) {
  throw new Error(`duplicate factor or subfactor: ${factor}`)
}
mapping.set(factor, {value: 0.5 + i*(1 + padding) + offset})
```
which keeps the common code one level higher.